# #7.9 findById Test

## findByID() 테스트

### findByID() 코드 수정

users.service.ts

```ts
try {
  const user = await this.users.findOneOrFail({ id }); // throw exception
  return {
    ok: true,
    user,
  };
} catch (error) {
  return {
    ok: false,
    error: '사용자를 찾을 수 없습니다.',
  };
}
```

- `findOne() `대신 `findOrFail()` 사용
  - 찾지 못했을 경우 에러를 발생시킨다 (-> `catch`)

### mockRepository에 findOrFail() 추가

users.service.spec.ts

```ts
const mockRepository = () => ({
  findOne: jest.fn(), // creates a mock function
  save: jest.fn(),
  create: jest.fn(),
  findOneOrFail: jest.fn(),
});
```

- 새로 사용되는 함수이므로 추가해야 된다.

### findById() 테스트 코드 작성

users.service.spec.ts

```ts
describe('findById', () => {
  const findByIdArgs = { id: 1 };
  it('should find an existing user', async () => {
    usersRepository.findOneOrFail.mockResolvedValue(findByIdArgs);
    const result = await service.findById(1);
    expect(result).toEqual({ ok: true, user: findByIdArgs });
  });
  it('should fail if user not found ', async () => {
    usersRepository.findOneOrFail.mockRejectedValue(new Error());
    const result = await service.findById(1);
    expect(result).toEqual({
      ok: false,
      error: '사용자를 찾을 수 없습니다.',
    });
  });
});
```
